cpp/src/datetime/datetime_ops.cu

```diff
@@ -263,29 +268,66 @@ std::unique_ptr<column> add_calendrical_months(column_view const& timestamp_colu
                                                rmm::mr::device_memory_resource* mr)
 {
   CUDF_EXPECTS(is_timestamp(timestamp_column.type()), "Column type should be timestamp");
-  CUDF_EXPECTS(months_column.type() == data_type{type_id::INT16},
-               "Months column type should be INT16");
+  CUDF_EXPECTS(
+    months_column.type().id() == type_id::INT16 or months_column.type().id() == type_id::INT32,
+    "Months column type should be INT16 or INT32.");
   CUDF_EXPECTS(timestamp_column.size() == months_column.size(),
                "Timestamp and months column should be of the same size");
-  auto size            = timestamp_column.size();
-  auto output_col_type = timestamp_column.type();
-
-  // Return an empty column if source column is empty
-  if (size == 0) return make_empty_column(output_col_type);
-
-  auto output_col_mask =
-    cudf::detail::bitmask_and(table_view({timestamp_column, months_column}), stream, mr);
-  auto output = make_fixed_width_column(
-    output_col_type, size, std::move(output_col_mask), cudf::UNKNOWN_NULL_COUNT, stream, mr);
-
-  auto launch = add_calendrical_months_functor{
-    timestamp_column, months_column, static_cast<mutable_column_view>(*output)};
-
-  type_dispatcher(timestamp_column.type(), launch, stream);
 
+  auto output = months_column.type() == data_type{type_id::INT16}
+                  ? type_dispatcher(timestamp_column.type(),
+                                    add_calendrical_months_functor{},
+                                    timestamp_column,
+                                    months_column.begin<int16_t>(),
+                                    stream,
+                                    mr)
+                  : type_dispatcher(timestamp_column.type(),
+                                    add_calendrical_months_functor{},
+                                    timestamp_column,
+                                    months_column.begin<int32_t>(),
+                                    stream,
+                                    mr);
```

Comment:
Just input a `months_begin_iter` of type `indexalator`.
```
auto const months_begin_iter = cudf::detail::indexalator_factory::make_input_iterator(months_column);
```

Reply:
Nice! This is really clean.